contracts/extension/plugin/Map.sol

```diff
@@ -3,61 +3,103 @@ pragma solidity ^0.8.0;
+    }
+
+    function _setPlugin(Plugin memory _plugin) internal {
+        require(allSelectors.add(bytes32(_plugin.selector)), "REGISTERED");
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
+
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
 
-        emit ExtensionRegistered(_selector, _extension);
+        emit PluginRegistered(_plugin.selector, _plugin.pluginAddress);
     }
 
-    function getExtensionForFunction(bytes4 _selector) external view returns (address) {
-        address ext = extension[_selector];
-        require(ext != address(0), "No extension available for selector.");
+    function _updatePlugin(Plugin memory _plugin) internal {
+        address currentExtension = getPluginForFunction(_plugin.selector);
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
 
-        return ext;
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
+        selectorsForPlugin[currentExtension].remove(bytes32(_plugin.selector));
+
+        emit PluginUpdated(_plugin.selector, currentExtension, _plugin.pluginAddress);
+    }
+
+    function _removePlugin(bytes4 _selector) internal {
+        address currentExtension = getPluginForFunction(_selector);
+
+        delete pluginForSelector[_selector];
+        allSelectors.remove(_selector);
+        selectorsForPlugin[currentExtension].remove(bytes32(_selector));
+
+        emit PluginRemoved(_selector, currentExtension);
     }
 
-    function getAllFunctionsOfExtension(address _extension) external view returns (bytes4[] memory registered) {
-        uint256 len = selectorsForExtension[_extension].length();
+    function getPluginForFunction(bytes4 _selector) public view returns (address) {
+        address _pluginAddress = pluginForSelector[_selector].pluginAddress;
+        require(_pluginAddress != address(0), "No plugin available for selector.");
+
+        return _pluginAddress;
+    }
+
+    function getAllFunctionsOfPlugin(address _extension) external view returns (bytes4[] memory registered) {
+        uint256 len = selectorsForPlugin[_extension].length();
         registered = new bytes4[](len);
 
         for (uint256 i = 0; i < len; i += 1) {
-            registered[i] = bytes4(selectorsForExtension[_extension].at(i));
+            registered[i] = bytes4(selectorsForPlugin[_extension].at(i));
         }
     }
 
-    function getAllRegistered() external view returns (ExtensionMap[] memory functionExtensionPairs) {
+    function getAllRegistered() external view returns (Plugin[] memory functionExtensionPairs) {
         uint256 len = allSelectors.length();
-        functionExtensionPairs = new ExtensionMap[](len);
+        functionExtensionPairs = new Plugin[](len);
```

Comment:
Replace all `extension` vocabulary with `plugin` vocabulary, even for variables e.g. `functionExtensionPairs`.